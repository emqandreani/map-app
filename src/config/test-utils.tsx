import { StyleSystemProvider } from "@architecture-it/stylesystem";
import { configureStore, PreloadedState } from "@reduxjs/toolkit";
import { render, renderHook } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import getTheme from "@architecture-it/stylesystem/Theme";
import React, { PropsWithChildren } from "react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";

import { ModuleContext } from "../store/hooks";
import { RootState, store } from "../store/store";
import { reducers } from "../store/reducers";
// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: typeof store;
  basename?: string;
}

interface WithProvidersProps {
  basename?: string;
  store: typeof store;
}

const colors = {
  primary: "#d71920",
  colorText: "#616161",
  error: "#d71920",
  secondary: "#b6040b",
};

export function WithProviders({ basename, store }: WithProvidersProps) {
  return function Wrapper({ children }: PropsWithChildren<{}>) {
    return (
      <BrowserRouter basename={basename}>
        <StyleSystemProvider theme={getTheme(false, colors)}>
          <Provider context={ModuleContext} store={store}>
            {children}
          </Provider>
        </StyleSystemProvider>
      </BrowserRouter>
    );
  };
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducers, preloadedState }),
    // Basename might be "/" for modules
    basename = "/",
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: WithProviders({ basename, store }), ...renderOptions }),
  };
}

export function renderHookWithProviders<Result, Props>(
  hook: (_initialProps: Props) => Result,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: reducers, preloadedState }),
    // Basename might be "/" for modules
    basename = "/",
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...renderHook(hook, { wrapper: WithProviders({ basename, store }), ...renderOptions }),
  };
}
