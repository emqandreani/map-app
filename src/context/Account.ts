"use client";
import React from "react";
import { IUseUser } from "@architecture-it/react-auth/use-user";

import { IPermission } from "../domain/Permission";

/**
 * Context to keep and share account info inside microfontend
 */
export const UserContext = React.createContext<IUseUser<IPermission> | undefined>(undefined);

export const useUserContext = () => React.useContext(UserContext);

export const DEFAULT_USER: IUseUser<IPermission> = {
  username: "Lmessi",
  email: "lmessi@ejemplo.com",
  isFromAAD: false,
  name: "Lionel",
};
