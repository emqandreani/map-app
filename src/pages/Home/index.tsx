import { Box, Typography } from "@mui/material";

import styles from "./styles.module.scss";

export default function Home() {
  return (
    <Box className={styles.page} data-testid="home-page">
      <Typography className={styles.text} data-testid="text" variant="h2">
        Map
      </Typography>
    </Box>
  );
}
