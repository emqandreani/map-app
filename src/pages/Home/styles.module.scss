@import "../../layout/general";

.page {
  @extend %page-styles;

  flex-direction: column;
  justify-content: center;
  gap: var(--spacing-6);

  .text {
    font-size: var(--title-3);
    color: var(--primary);

    .span {
      color: var(--color-text-base);
    }
  }
}
