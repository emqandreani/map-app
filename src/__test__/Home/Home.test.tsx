import { fireEvent } from "@testing-library/react";
import { useNavigate } from "react-router-dom";

import { renderWithProviders } from "../../config/test-utils";
import { useUserContext, DEFAULT_USER } from "../../context/Account";
import Home from "../../pages/Home";

jest.mock("react-router-dom", () => {
  const actual: object = jest.requireActual("react-router-dom");

  return {
    ...actual,
    useNavigate: jest.fn().mockImplementation(() => jest.fn()),
  };
});

jest.mock("../../context/Account", () => {
  const actual: object = jest.requireActual("../../context/Account");

  return {
    ...actual,
    useUserContext: jest
      .fn()
      .mockImplementation(() => DEFAULT_USER)
      .mockImplementationOnce(() => undefined),
  };
});

const useNavigateMocked = useNavigate as jest.Mock;

describe("<Home />", () => {
  test("Should render without crash if not authenticated", () => {
    const { getByTestId } = renderWithProviders(<Home />);

    const HomePage = getByTestId("home-page");
    const text = getByTestId("text");

    expect(HomePage).toMatchSnapshot();
    expect(text.textContent).toBe("Por favor inicia sesión para ver la información de usuario.");
  });
  test("Should render welcome text if authenticated and account exist", () => {
    const { getByTestId } = renderWithProviders(<Home />);

    const HomePage = getByTestId("home-page");
    const text = getByTestId("text");

    expect(HomePage).toMatchSnapshot();
    expect(text.textContent).toBe(`Hola ${DEFAULT_USER.name}!`);
  });
  test("Should handle onclick", () => {
    const { getByTestId } = renderWithProviders(<Home />);

    const button = getByTestId("button");

    fireEvent.click(button);

    expect(useNavigateMocked).toHaveBeenCalled();
    // expect(useNavigateMocked).toHaveBeenCalledWith("usuarios");
  });
});
