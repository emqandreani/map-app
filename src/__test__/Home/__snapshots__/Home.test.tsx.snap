// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Should render welcome text if authenticated and account exist 1`] = `
<div
  class="page MuiBox-root css-0"
  data-testid="home-page"
>
  <h1
    class="MuiTypography-root MuiTypography-h1 text css-aki82s-MuiTypography-root"
    data-testid="text"
  >
    <span
      class="span"
    >
      Hola
    </span>
     
    Lionel
    !
  </h1>
  <button
    class="MuiButtonBase-root MuiButton-root StyleSystem__Button_root MuiButton-contained StyleSystem__Button_contained MuiButton-containedPrimary MuiButton-sizeMedium StyleSystem__Button_sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root StyleSystem__Button_root MuiButton-contained StyleSystem__Button_contained MuiButton-containedPrimary MuiButton-sizeMedium StyleSystem__Button_sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation css-qz5d3w-MuiButtonBase-root-MuiButton-root"
    data-testid="button"
    tabindex="0"
    type="button"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 css-rtjrfq-MuiTypography-root"
    >
      Ver usuarios
    </h6>
    <span
      class="MuiTouchRipple-root css-1r0r0n5-MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`<Home /> Should render without crash if not authenticated 1`] = `
<div
  class="page MuiBox-root css-0"
  data-testid="home-page"
>
  <h2
    class="MuiTypography-root MuiTypography-h2 text css-u5kjcz-MuiTypography-root"
    data-testid="text"
  >
    Por favor inicia sesión para ver la información de usuario.
  </h2>
</div>
`;
